### PR TITLE
Correcting Quadrilateral3D4N jacobian and area

### PR DESCRIPTION
**Pull request by loumalouomega**
Maybe not 100% solved, some tests I am doing are not working yet, for some reason.

### CODE REVIEW

File: kratos/geometries/quadrilateral_3d_4.h [hunk 1]
```diff
@@ -408,65 +407,73 @@ template<class TPointType> class Quadrilateral3D4
      */
     double Area() const override
     {
-        // Old
-
-        /*Vector d = this->Points()[2] - this->Points()[0];
-        return( sqrt( d[0]*d[0] + d[1]*d[1] + d[2]*d[2] ) );*/
+        // Finite element way
+        Vector temp;
+        DeterminantOfJacobian( temp, GeometryData::GI_GAUSS_3 );
```
Comment (pooyan-dadvand) [inline kratos/geometries/quadrilateral_3d_4.h]: I'm not sure if creating the Vector is not harmful for efficiency. I think you can move the jacobian implemenation to a private method passing an array_1d with correct size and then use it in both methods.

File: kratos/geometries/quadrilateral_3d_4.h [hunk 2]
```diff
@@ -485,33 +492,18 @@ template<class TPointType> class Quadrilateral3D4
      */
     double DomainSize() const override
     {
-		// Old
+        // Old
```
Comment (pooyan-dadvand) [inline kratos/geometries/quadrilateral_3d_4.h]: I would remove the comments....

File: kratos/geometries/quadrilateral_3d_4.h [hunk 3]
```diff
@@ -608,15 +599,15 @@ template<class TPointType> class Quadrilateral3D4
...
 
             if ( norm_2( DeltaXi ) > 300 )
             {
                 res[0] = 0.0;
                 res[1] = 0.0;
+            #ifdef KRATOS_DEBUG 
                 std::cout << "detJ =" << det_j << "DeltaX = " << DeltaXi << " stopping calculation and assigning the baricenter" << std::endl;
+            #endif  
                 break;
-                //KRATOS_ERROR << "Computation of local coordinates failed at iteration" << k << std:endl;
```
Comment (pooyan-dadvand) [inline kratos/geometries/quadrilateral_3d_4.h]: This is changing the behavior and I'm not sure if is a good idea to return a zero with just a small hint....

File: kratos/geometries/quadrilateral_3d_4.h [hunk 4]
```diff
@@ -649,16 +640,17 @@ template<class TPointType> class Quadrilateral3D4
...
+        JacobiansType& rResult,
+        IntegrationMethod ThisMethod 
+        ) const override
     {
-        //getting derivatives of shape functions
+        // Getting derivatives of shape functions
         ShapeFunctionsGradientsType shape_functions_gradients =
-            CalculateShapeFunctionsIntegrationPointsLocalGradients( ThisMethod );
-        //getting values of shape functions
+        CalculateShapeFunctionsIntegrationPointsLocalGradients( ThisMethod );
```
Comment (pooyan-dadvand) [inline kratos/geometries/quadrilateral_3d_4.h]: You should not use the private calculate method. You have the calculated values in geometry data and can be used via ShapeFunctionsIntegrationPointsLocalGradients which return a reference. So no copy is needed!!

File: kratos/geometries/quadrilateral_3d_4.h [hunk 5]
```diff
@@ -649,16 +640,17 @@ template<class TPointType> class Quadrilateral3D4
...
+        // Getting values of shape functions
         Matrix shape_functions_values =
-            CalculateShapeFunctionsIntegrationPointsValues( ThisMethod );
-        //workaround by riccardo...
+        CalculateShapeFunctionsIntegrationPointsValues( ThisMethod );
```
Comment (pooyan-dadvand) [inline kratos/geometries/quadrilateral_3d_4.h:576]: Avoid Matrix generation by calling the ShapeFunctionsValues method

File: kratos/geometries/quadrilateral_3d_4.h [hunk 6]
```diff
@@ -715,17 +707,17 @@ template<class TPointType> class Quadrilateral3D4
...
+        JacobiansType& rResult,
+        IntegrationMethod ThisMethod,
+        Matrix & DeltaPosition 
+        ) const override
     {
-        //getting derivatives of shape functions
+        // Getting derivatives of shape functions
         ShapeFunctionsGradientsType shape_functions_gradients =
-            CalculateShapeFunctionsIntegrationPointsLocalGradients( ThisMethod );
+        CalculateShapeFunctionsIntegrationPointsLocalGradients( ThisMethod );
```
Comment (pooyan-dadvand) [inline kratos/geometries/quadrilateral_3d_4.h]: Same....

File: kratos/geometries/quadrilateral_3d_4.h [hunk 7]
```diff
@@ -715,17 +707,17 @@ template<class TPointType> class Quadrilateral3D4
...
         //getting values of shape functions
-        Matrix shape_functions_values =
-            CalculateShapeFunctionsIntegrationPointsValues( ThisMethod );
-        //workaround by riccardo...
+        Matrix shape_functions_values = CalculateShapeFunctionsIntegrationPointsValues( ThisMethod );
```
Comment (pooyan-dadvand) [inline kratos/geometries/quadrilateral_3d_4.h]: again....

File: kratos/geometries/quadrilateral_3d_4.h [hunk 8]
```diff
@@ -784,41 +776,33 @@ template<class TPointType> class Quadrilateral3D4
...
-            shape_functions_gradients( IntegrationPointIndex );
-        //values of shape functions in integration points
-//         Vector ShapeFunctionsValuesInIntegrationPoint = ZeroVector( 4 );
-//         /*vector<double>*/
-//         ShapeFunctionsValuesInIntegrationPoint = row(
-//                 CalculateShapeFunctionsIntegrationPointsValues( ThisMethod ),
-//                 IntegrationPointIndex );
-
+        // Derivatives of shape functions
+        ShapeFunctionsGradientsType shape_functions_gradients = CalculateShapeFunctionsIntegrationPointsLocalGradients( ThisMethod );
```
Comment (pooyan-dadvand) [inline kratos/geometries/quadrilateral_3d_4.h]: Again....

Comment (pooyan-dadvand) [reply kratos/geometries/quadrilateral_3d_4.h]: If you get it correctly for your method you don't need to access the index again below.

File: kratos/geometries/quadrilateral_3d_4.h [hunk 9]
```diff
@@ -784,41 +776,33 @@ template<class TPointType> class Quadrilateral3D4
...
+        
         //Elements of jacobian matrix (e.g. J(1,1) = dX1/dXi1)
         //loop over all nodes
-
         for ( unsigned int i = 0; i < this->PointsNumber(); i++ )
         {
             rResult( 0, 0 ) +=
-                ( this->GetPoint( i ).X() ) * ( ShapeFunctionsGradientInIntegrationPoint( i, 0 ) );
+                ( this->GetPoint( i ).X() ) * ( shape_functions_gradients[IntegrationPointIndex]( i, 0 ) );
```
Comment (pooyan-dadvand) [inline kratos/geometries/quadrilateral_3d_4.h]: It is not needed if you get the right reference

File: kratos/geometries/quadrilateral_3d_4.h [hunk 10]
```diff
@@ -847,8 +831,7 @@ template<class TPointType> class Quadrilateral3D4
 
         //derivatives of shape functions
         Matrix shape_functions_gradients;
```
Comment (pooyan-dadvand) [inline kratos/geometries/quadrilateral_3d_4.h]: Matrix&

File: kratos/geometries/quadrilateral_3d_4.h [hunk 11]
```diff
@@ -925,16 +909,18 @@ template<class TPointType> class Quadrilateral3D4
...
-                                          IntegrationMethod ThisMethod ) const override
+    double DeterminantOfJacobian( 
+        IndexType IntegrationPointIndex,
+        IntegrationMethod ThisMethod 
+        ) const override
     {
         Matrix jacobian ( 3, 2 );
          
-        this->Jacobian( jacobian, IntegrationPointIndex, ThisMethod);
+        Jacobian( jacobian, IntegrationPointIndex, ThisMethod);
```
Comment (pooyan-dadvand) [inline kratos/geometries/quadrilateral_3d_4.h]: Are you sure!?!

File: kratos/geometries/quadrilateral_3d_4.h [hunk 12]
```diff
@@ -925,16 +909,18 @@ template<class TPointType> class Quadrilateral3D4
...
             
-        const double detJ = std::pow(jacobian(0,1),2)*(std::pow(jacobian(1,0),2)+std::pow(jacobian(2,0),2))+std::pow(jacobian(1,1)*jacobian(2,0)-jacobian(1,0)*jacobian(2,1),2)-2*jacobian(0,0)*jacobian(0,1)*(jacobian(1,0)*jacobian(1,1)+jacobian(2,0)*jacobian(2,1))+std::pow(jacobian(0,0),2)*(std::pow(jacobian(1,1),2)+std::pow(jacobian(2,1),2));
+        const double detJ = std::sqrt(std::pow(jacobian(0,1),2.0)*(std::pow(jacobian(1,0),2.0) + std::pow(jacobian(2,0),2.0)) + std::pow(jacobian(1,1)*jacobian(2,0) - jacobian(1,0)*jacobian(2,1),2.0) - 2.0*jacobian(0,0)*jacobian(0,1)*(jacobian(1,0)*jacobian(1,1) + jacobian(2,0)*jacobian(2,1)) + std::pow(jacobian(0,0),2.0)*(std::pow(jacobian(1,1),2.0) + std::pow(jacobian(2,1),2.0)));
```
Comment (pooyan-dadvand) [inline kratos/geometries/quadrilateral_3d_4.h]: I would separate the sqrt and put a debug check if detJ is negative.....

**Review by pooyan-dadvand — CHANGES_REQUESTED** [review]
I have many efficiency issues. Many of them are historical...

File: kratos/geometries/quadrilateral_3d_4.h (hunk 1 above)
Comment (loumalouomega) [reply kratos/geometries/quadrilateral_3d_4.h]: Yeah, that can work

File: kratos/geometries/quadrilateral_3d_4.h (hunk 2 above)
Comment (loumalouomega) [reply kratos/geometries/quadrilateral_3d_4.h]: old comments

File: kratos/geometries/quadrilateral_3d_4.h (hunk 3 above)
Comment (loumalouomega) [reply kratos/geometries/quadrilateral_3d_4.h]: Ok

File: kratos/geometries/quadrilateral_3d_4.h (hunk 4 above)
Comment (loumalouomega) [reply kratos/geometries/quadrilateral_3d_4.h]: Ok

File: kratos/geometries/quadrilateral_3d_4.h (hunk 5 above)
Comment (loumalouomega) [reply kratos/geometries/quadrilateral_3d_4.h:576]: Ok

File: kratos/geometries/quadrilateral_3d_4.h (hunk 6 above)
Comment (loumalouomega) [reply kratos/geometries/quadrilateral_3d_4.h]: Ok

File: kratos/geometries/quadrilateral_3d_4.h (hunk 7 above)
Comment (loumalouomega) [reply kratos/geometries/quadrilateral_3d_4.h]: Ok

File: kratos/geometries/quadrilateral_3d_4.h (hunk 12 above)
Comment (loumalouomega) [reply kratos/geometries/quadrilateral_3d_4.h]: Ok

File: kratos/geometries/quadrilateral_3d_4.h (hunk 4 above)
Comment (loumalouomega) [reply kratos/geometries/quadrilateral_3d_4.h]: Should I change this in all geometries?

File: kratos/geometries/quadrilateral_3d_4.h (hunk 8 above)
Comment (loumalouomega) [reply kratos/geometries/quadrilateral_3d_4.h]: I don't understand

File: kratos/geometries/quadrilateral_3d_4.h (hunk 9 above)
Comment (loumalouomega) [reply kratos/geometries/quadrilateral_3d_4.h]: Ok

File: kratos/geometries/quadrilateral_3d_4.h (hunk 10 above)
Comment (loumalouomega) [reply kratos/geometries/quadrilateral_3d_4.h]: Doesn't work

File: kratos/geometries/quadrilateral_3d_4.h (hunk 11 above)
Comment (loumalouomega) [reply kratos/geometries/quadrilateral_3d_4.h]: It is the same

File: kratos/geometries/quadrilateral_3d_4.h [hunk 13]
```diff
@@ -647,7 +570,7 @@ template<class TPointType> class Quadrilateral3D4
     {
         // Getting derivatives of shape functions
         ShapeFunctionsGradientsType shape_functions_gradients =
```
Comment (pooyan-dadvand) [inline kratos/geometries/quadrilateral_3d_4.h]: You need to use reference here to make it efficient: `ShapeFunctionsGradientsType & shape_functions_gradients =`

File: kratos/geometries/quadrilateral_3d_4.h [hunk 14]
```diff
@@ -715,8 +638,8 @@ template<class TPointType> class Quadrilateral3D4
...
```
Comment (pooyan-dadvand) [inline kratos/geometries/quadrilateral_3d_4.h]: Again

File: kratos/geometries/quadrilateral_3d_4.h [hunk 15]
```diff
@@ -647,7 +570,7 @@ template<class TPointType> class Quadrilateral3D4
     {
         // Getting derivatives of shape functions
         ShapeFunctionsGradientsType shape_functions_gradients =
-        CalculateShapeFunctionsIntegrationPointsLocalGradients( ThisMethod );
+        msGeometryData.ShapeFunctionsLocalGradients( ThisMethod );
         // Getting values of shape functions
         Matrix shape_functions_values =
```
Comment (pooyan-dadvand) [inline kratos/geometries/quadrilateral_3d_4.h]: use a reference please

File: kratos/geometries/quadrilateral_3d_4.h [hunk 16]
```diff
@@ -715,8 +638,8 @@ template<class TPointType> class Quadrilateral3D4
     {
         // Getting derivatives of shape functions
         ShapeFunctionsGradientsType shape_functions_gradients =
-        CalculateShapeFunctionsIntegrationPointsLocalGradients( ThisMethod );
-        //getting values of shape functions
+        msGeometryData.ShapeFunctionsLocalGradients( ThisMethod );
+        // Getting values of shape functions
         Matrix shape_functions_values = CalculateShapeFunctionsIntegrationPointsValues( ThisMethod );
```
Comment (pooyan-dadvand) [inline kratos/geometries/quadrilateral_3d_4.h]: more...

File: kratos/geometries/quadrilateral_3d_4.h (hunk 10 above)
Comment (pooyan-dadvand) [reply kratos/geometries/quadrilateral_3d_4.h]: you should create the reference when you are assigning the gradients.

File: kratos/geometries/quadrilateral_3d_4.h [hunk 17]
```diff
@@ -952,11 +879,13 @@ template<class TPointType> class Quadrilateral3D4
...
         Matrix jacobian ( 3, 2 );
          
-        Jacobian( jacobian, rPoint);
+        this->Jacobian( jacobian, rPoint);
         
-        const double detJ = std::sqrt(std::pow(jacobian(0,1),2.0)*(std::pow(jacobian(1,0),2.0) + std::pow(jacobian(2,0),2.0)) + std::pow(jacobian(1,1)*jacobian(2,0) - jacobian(1,0)*jacobian(2,1),2.0) - 2.0*jacobian(0,0)*jacobian(0,1)*(jacobian(1,0)*jacobian(1,1) + jacobian(2,0)*jacobian(2,1)) + std::pow(jacobian(0,0),2.0)*(std::pow(jacobian(1,1),2.0) + std::pow(jacobian(2,1),2.0)));
+        const double detJ = std::pow(jacobian(0,1),2.0)*(std::pow(jacobian(1,0),2.0) + std::pow(jacobian(2,0),2.0)) + std::pow(jacobian(1,1)*jacobian(2,0) - jacobian(1,0)*jacobian(2,1),2.0) - 2.0*jacobian(0,0)*jacobian(0,1)*(jacobian(1,0)*jacobian(1,1) + jacobian(2,0)*jacobian(2,1)) + std::pow(jacobian(0,0),2.0)*(std::pow(jacobian(1,1),2.0) + std::pow(jacobian(2,1),2.0));
         
-        return detJ;
+        if (detJ < 0.0) KRATOS_ERROR << "WARNING::NEGATIVE VALUE: NOT POSSIBLE TO EVALUATE THE JACOBIAN DETERMINANT" << std::endl;
```
Comment (pooyan-dadvand) [inline kratos/geometries/quadrilateral_3d_4.h]: again

File: kratos/geometries/quadrilateral_3d_4.h [hunk 18]
```diff
@@ -1230,8 +1158,8 @@ template<class TPointType> class Quadrilateral3D4
             rResult.swap( temp );
         }
 
-        //calculating the local gradients
-        ShapeFunctionsGradientsType locG = CalculateShapeFunctionsIntegrationPointsLocalGradients( ThisMethod );
+        // Calculating the local gradients
+        ShapeFunctionsGradientsType locG = msGeometryData.ShapeFunctionsLocalGradients( ThisMethod );
```
Comment (pooyan-dadvand) [inline kratos/geometries/quadrilateral_3d_4.h]: reference

File: kratos/geometries/quadrilateral_3d_4.h [hunk 19]
```diff
@@ -1315,7 +1243,7 @@ template<class TPointType> class Quadrilateral3D4
         IntegrationMethod ThisMethod )
     {
         ShapeFunctionsGradientsType localGradients
```
Comment (pooyan-dadvand) [inline kratos/geometries/quadrilateral_3d_4.h]: reference

File: kratos/geometries/quadrilateral_3d_4.h [hunk 20]
```diff
@@ -1315,7 +1243,7 @@ template<class TPointType> class Quadrilateral3D4
         IntegrationMethod ThisMethod )
     {
         ShapeFunctionsGradientsType localGradients
-        = CalculateShapeFunctionsIntegrationPointsLocalGradients( ThisMethod );
+        = msGeometryData.ShapeFunctionsLocalGradients( ThisMethod );
         const int integration_points_number
         = msGeometryData.IntegrationPointsNumber( ThisMethod );
         ShapeFunctionsGradientsType Result( integration_points_number );
```
Comment (pooyan-dadvand) [inline kratos/geometries/quadrilateral_3d_4.h]: reference and initialize it at place if is possible

File: kratos/geometries/quadrilateral_3d_4.h [hunk 21]
```diff
@@ -1336,7 +1264,7 @@ template<class TPointType> class Quadrilateral3D4
     {
         IntegrationMethod ThisMethod = msGeometryData.DefaultIntegrationMethod();
         ShapeFunctionsGradientsType localGradients
```
Comment (pooyan-dadvand) [inline kratos/geometries/quadrilateral_3d_4.h]: more

File: kratos/geometries/quadrilateral_3d_4.h [hunk 22]
```diff
@@ -1336,7 +1264,7 @@ template<class TPointType> class Quadrilateral3D4
     {
         IntegrationMethod ThisMethod = msGeometryData.DefaultIntegrationMethod();
         ShapeFunctionsGradientsType localGradients
-        = CalculateShapeFunctionsIntegrationPointsLocalGradients( ThisMethod );
+        = msGeometryData.ShapeFunctionsLocalGradients( ThisMethod );
         const int integration_points_number
         = msGeometryData.IntegrationPointsNumber( ThisMethod );
         ShapeFunctionsGradientsType Result( integration_points_number );
```
Comment (pooyan-dadvand) [inline kratos/geometries/quadrilateral_3d_4.h]: more

**Review by pooyan-dadvand — CHANGES_REQUESTED** [review]
@loumalouomega Thanks for the changes. However there are few more to do. And please revise the entire code for similar "historical" efficiency problems.

File: kratos/geometries/quadrilateral_3d_4.h (hunk 13 above)
Comment (loumalouomega) [reply kratos/geometries/quadrilateral_3d_4.h]: I can't, it is const, so I cannot use it (doesn't compile)

File: kratos/geometries/quadrilateral_3d_4.h (hunk 15 above)
Comment (loumalouomega) [reply kratos/geometries/quadrilateral_3d_4.h]: Idem

File: kratos/geometries/quadrilateral_3d_4.h (hunk 14 above)
Comment (loumalouomega) [reply kratos/geometries/quadrilateral_3d_4.h]: Idem

File: kratos/geometries/quadrilateral_3d_4.h (hunk 16 above)
Comment (loumalouomega) [reply kratos/geometries/quadrilateral_3d_4.h]: Idem

File: kratos/geometries/quadrilateral_3d_4.h (hunk 17 above)
Comment (loumalouomega) [reply kratos/geometries/quadrilateral_3d_4.h]: Where?

**Review by pooyan-dadvand — APPROVED** [review]
Good Job! 👍

**Comment by pooyan-dadvand** [comment]
By the way, is it 100% working?

**Comment by loumalouomega** [comment]
I am doing some extra debugging, apparently I am having some IsInside() false positive, which give a  poor convergence because random triangles appear in the exact integration

**Comment by pooyan-dadvand** [comment]
I see that you are using the same tolerance everywhere which is a good move! Is it improving the results?

**Comment by loumalouomega** [comment]
I needed to restore one tolerance, I was having many false positive

**Pull request merged**